query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/PinotClientFactory.java

```diff
@@ -102,5 +108,47 @@ private PreparedStatement buildPreparedStatement(String statement, Params params
           .forEach((i, b) -> preparedStatement.setString(i, Hex.encodeHexString(b.toByteArray())));
       return preparedStatement;
     }
+
+    @VisibleForTesting
+    /*
+     * Pinot PreparedStatement creates invalid query if one of the parameters has '?' in its value.
+     * Sample pinot query : select * from table where team in (?, ?, ?).. Now say parameters are:
+     * 'abc', 'pqr with (?)' and 'xyz'..
+     *
+     * Now, on executing PreparedStatement:fillStatementWithParameters method on this will return
+     * select * from table where team in ('abc', 'pqr with ('xyz')', ?) -- which is clearly wrong
+     * (what we wanted was select * from table where team in ('abc', 'pqr with (?)', 'xyz'))..
+     *
+     * The reason is the usage of replaceFirst iteration in the pinot PreparedStatement method..
+     *
+     * Hence written this custom method to resolve the query statement rather than relying on Pinot's
+     * library method.
+     * This is a temporary fix and can be reverted when the Pinot issue gets resolved
+     * Raised an issue in incubator-pinot github repo: apache/incubator-pinot#6834
+     */
+    String resolveStatement(String query, Params params) {
```

Comment:
What happens in this method when the number of `?` don't match the number of params?

Reply:
if number of params are more, we drop the extras and if it's less, we replace `?` with empty string.. the loop takes care of this..